File: apps/client-macos/Sources/TimeTrack/Storage/BufferStore.swift
import Foundation

/// The kind of a buffered record — used to route on drain (SyncEngine syncs each kind to its own endpoint).
enum BufferKind: String {
    case timeEntry
    case idleEvent
}

/// PRD §7.5 — durable, file-backed offline write-buffer. One atomic file per record under
/// Application Support; the filename `<createdAtMillis>__<kind>__<uuidv7>.json` carries FIFO order,
/// routing, and identity, so listing the directory yields all three WITHOUT reading contents. The
/// file's content IS the raw payload the API upserts on (idempotent on the UUIDv7). Hand-rolled —
/// no SQLite dependency (CLAUDE.md §2); durability comes from write-temp-then-rename plus a startup
/// sweep of any `.tmp-*` left by a crash between write and rename. File-per-record gives natural
/// isolation: concurrent enqueue (main) and take/remove (sync task) touch different files.
final class BufferStore {
    static let shared = BufferStore(directory: BufferStore.defaultDirectory())

    private let directory: URL
    private let clock: () -> Date

    init(directory: URL, clock: @escaping () -> Date = Date.init) {
        self.directory = directory
        self.clock = clock
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        sweepTemporaries()
    }

    /// ~/Library/Application Support/<container>/buffer/ — per-install; see `AppInstall`.
    static func defaultDirectory() -> URL {
        AppInstall.supportDirectory("buffer")
    }

    /// Atomic enqueue: write to `.tmp-<id>`, then rename to the final `<millis>__<kind>__<id>.json`.
    func enqueue(id: String, kind: BufferKind, payload: Data) {
        let millis = Int64(clock().timeIntervalSince1970 * 1000)
        let dst = directory.appendingPathComponent("\(millis)__\(kind.rawValue)__\(id).json")
        let tmp = directory.appendingPathComponent(".tmp-\(id)")
        do {
            try payload.write(to: tmp, options: .atomic)
            try? FileManager.default.removeItem(at: dst)   // ids are unique; defensive
            try FileManager.default.moveItem(at: tmp, to: dst)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
        }
    }

    /// FIFO (createdAt asc) records of `kind`, up to `limit`.
    func take(kind: BufferKind, limit: Int) -> [(id: String, payload: Data)] {
        allRecords()
            .filter { $0.kind == kind.rawValue }
            .prefix(limit)
            .compactMap { rec in
                guard let data = try? Data(contentsOf: rec.url) else { return nil }
                return (id: rec.id, payload: data)
            }
    }

    func remove(id: String) {
        for rec in allRecords() where rec.id == id {
            try? FileManager.default.removeItem(at: rec.url)
        }
    }

    /// Drops records created before `now - maxAge`, bounding the buffer against records that never
    /// deliver. Both kinds are drained and removed on 2xx, so only stuck records ever age out.
    func prune(olderThan maxAge: TimeInterval) {
        let cutoff = Int64(clock().addingTimeInterval(-maxAge).timeIntervalSince1970 * 1000)
        for rec in allRecords() where rec.createdAtMillis < cutoff {
            try? FileManager.default.removeItem(at: rec.url)
        }
    }

    func clear() {
        for rec in allRecords() { try? FileManager.default.removeItem(at: rec.url) }
    }

    // MARK: - internals

    private struct Record { let url: URL; let createdAtMillis: Int64; let kind: String; let id: String }

    private func contents() -> [URL] {
        (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)) ?? []
    }

    /// How many records are still waiting to reach the server.
    ///
    /// Counted from the directory listing, not by reading any payload — the filename carries
    /// everything, which is the whole point of the naming scheme.
    func pendingCount() -> Int {
        allRecords().count
    }

    private func parse(_ url: URL) -> Record? {
        let name = url.lastPathComponent
        guard name.hasSuffix(".json") else { return nil }
        let parts = String(name.dropLast(5)).components(separatedBy: "__")  // strip ".json"
        guard parts.count == 3, let millis = Int64(parts[0]) else { return nil }
        return Record(url: url, createdAtMillis: millis, kind: parts[1], id: parts[2])
    }

    private func allRecords() -> [Record] {
        contents().compactMap(parse)
            .sorted { ($0.createdAtMillis, $0.id) < ($1.createdAtMillis, $1.id) }
    }

    private func sweepTemporaries() {
        for url in contents() where url.lastPathComponent.hasPrefix(".tmp-") {
            try? FileManager.default.removeItem(at: url)
        }
    }
}

extension BufferStore: TimeEntryBuffering {}

File: apps/client-windows/src/NiftyTimer/Storage/BufferStore.cs
using System.Globalization;

namespace NiftyTimer.Storage;

/// <summary>
/// The kind of a buffered record — used to route on drain (SyncEngine syncs each kind to its own
/// endpoint).
/// </summary>
public enum BufferKind
{
    TimeEntry,
    IdleEvent,
}

/// <summary>The buffer seam. Tests substitute a spy.</summary>
public interface ITimeEntryBuffer
{
    void Enqueue(string id, BufferKind kind, byte[] payload);
}

/// <summary>
/// PRD §7.5 — durable, file-backed offline write-buffer. One atomic file per record under
/// %LOCALAPPDATA%; the filename <c>&lt;createdAtMillis&gt;__&lt;kind&gt;__&lt;uuidv7&gt;.json</c>
/// carries FIFO order, routing, and identity, so listing the directory yields all three WITHOUT
/// reading contents. The file's content IS the raw payload the API upserts on (idempotent on the
/// UUIDv7). Hand-rolled — no SQLite dependency (CLAUDE.md §2); durability comes from
/// write-temp-then-rename plus a startup sweep of any <c>.tmp-*</c> left by a crash between write
/// and rename. File-per-record gives natural isolation: concurrent enqueue (UI thread) and
/// take/remove (sync task) touch different files.
/// </summary>
public sealed class BufferStore : ITimeEntryBuffer
{
    private readonly string _directory;
    private readonly Func<DateTimeOffset> _clock;

    public BufferStore(string directory, Func<DateTimeOffset>? clock = null)
    {
        _directory = directory;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        Directory.CreateDirectory(directory);
        SweepTemporaries();
    }

    /// <summary>Atomic enqueue: write to <c>.tmp-&lt;id&gt;</c>, then rename to the final name.</summary>
    public void Enqueue(string id, BufferKind kind, byte[] payload)
    {
        var millis = _clock().ToUnixTimeMilliseconds();
        var dst = Path.Combine(_directory, $"{millis}__{KindToken(kind)}__{id}.json");
        var tmp = Path.Combine(_directory, $".tmp-{id}");
        try
        {
            File.WriteAllBytes(tmp, payload);
            File.Move(tmp, dst, overwrite: true); // ids are unique; overwrite is defensive
        }
        catch (IOException)
        {
            TryDelete(tmp);
        }
        catch (UnauthorizedAccessException)
        {
            TryDelete(tmp);
        }
    }

    /// <summary>FIFO (createdAt asc) records of <paramref name="kind"/>, up to <paramref name="limit"/>.</summary>
    public IReadOnlyList<(string Id, byte[] Payload)> Take(BufferKind kind, int limit)
    {
        var token = KindToken(kind);
        var result = new List<(string, byte[])>();
        foreach (var rec in AllRecords())
        {
            if (result.Count >= limit)
            {
                break;
            }

            if (rec.Kind != token)
            {
                continue;
            }

            try
            {
                result.Add((rec.Id, File.ReadAllBytes(rec.Path)));
            }
            catch (IOException)
            {
                // Unreadable mid-drain (deleted by a concurrent pass): skip it.
            }
        }

        return result;
    }

    public void Remove(string id)
    {
        foreach (var rec in AllRecords())
        {
            if (rec.Id == id)
            {
                TryDelete(rec.Path);
            }
        }
    }

    /// <summary>
    /// Drops records created before <c>now - maxAge</c>, bounding the buffer against records that
    /// never deliver. Both kinds are drained and removed on 2xx, so only stuck records ever age
    /// out.
    /// </summary>
    public void Prune(TimeSpan maxAge)
    {
        var cutoff = _clock().Subtract(maxAge).ToUnixTimeMilliseconds();
        foreach (var rec in AllRecords())
        {
            if (rec.CreatedAtMillis < cutoff)
            {
                TryDelete(rec.Path);
            }
        }
    }

    public void Clear()
    {
        foreach (var rec in AllRecords())
        {
            TryDelete(rec.Path);
        }
    }

    /// <summary>
    /// How many records are still waiting to reach the server.
    ///
    /// Counted from the directory listing, not by reading any payload — the filename carries
    /// everything, which is the whole point of the naming scheme, and this runs on every menu
    /// open.
    /// </summary>
    public int PendingCount() => AllRecords().Count;

    internal static string KindToken(BufferKind kind) => kind switch
    {
        BufferKind.TimeEntry => "timeEntry",
        BufferKind.IdleEvent => "idleEvent",
        _ => throw new ArgumentOutOfRangeException(nameof(kind)),
    };

    private readonly record struct Record(string Path, long CreatedAtMillis, string Kind, string Id);

    private string[] Contents()
    {
        try
        {
            return Directory.GetFiles(_directory);
        }
        catch (DirectoryNotFoundException)
        {
            return [];
        }
    }

    private static Record? Parse(string path)
    {
        var name = Path.GetFileName(path);
        if (!name.EndsWith(".json", StringComparison.Ordinal))
        {
            return null;
        }

        var parts = name[..^5].Split("__");
        if (parts.Length != 3 ||
            !long.TryParse(parts[0], NumberStyles.Integer, CultureInfo.InvariantCulture, out var millis))
        {
            return null;
        }

        return new Record(path, millis, parts[1], parts[2]);
    }

    private List<Record> AllRecords()
    {
        var records = new List<Record>();
        foreach (var path in Contents())
        {
            if (Parse(path) is { } rec)
            {
                records.Add(rec);
            }
        }

        records.Sort(static (a, b) =>
        {
            var byTime = a.CreatedAtMillis.CompareTo(b.CreatedAtMillis);
            return byTime != 0 ? byTime : string.CompareOrdinal(a.Id, b.Id);
        });
        return records;
    }

    private void SweepTemporaries()
    {
        foreach (var path in Contents())
        {
            if (Path.GetFileName(path).StartsWith(".tmp-", StringComparison.Ordinal))
            {
                TryDelete(path);
            }
        }
    }

    private static void TryDelete(string path)
    {
        try
        {
            File.Delete(path);
        }
        catch (IOException)
        {
            // Another pass got there first, or the file is momentarily locked. Either way the
            // next cycle retries; a failed delete must never abort a drain.
        }
        catch (UnauthorizedAccessException)
        {
        }
    }
}
